File: problems-adapt/0401-0500/0419_most-isolated-node-under-a-reach-budget/solution_dijkstra_every_node.py
import heapq


class Solution:
    def findMostIsolated(self, n: int, edges: list[list[int]], budget: int) -> int:
        # Mirror each undirected edge both ways, so every node can run its own
        # Dijkstra over the adjacency list and pay only for real edges.
        adj = [[] for _ in range(n)]
        for a, b, w in edges:
            adj[a].append((b, w))
            adj[b].append((a, w))
        INF = float("inf")
        counts = [0] * n
        for src in range(n):
            # Dijkstra from src: with positive weights the smallest tentative pop
            # is already final, so every node settles exactly once.
            dist = [INF] * n
            dist[src] = 0
            heap = [(0, src)]
            while heap:
                d, u = heapq.heappop(heap)
                # Stale-entry guard: skip outdated heap records.
                if d > dist[u]:
                    continue
                for v, w in adj[u]:
                    nd = d + w
                    # Relax only when the route strictly improves.
                    if nd < dist[v]:
                        dist[v] = nd
                        heapq.heappush(heap, (nd, v))
            counts[src] = sum(1 for v in range(n) if v != src and dist[v] <= budget)
        # Ascending scan with a strictly-smaller count (or equal count at a
        # larger index) implements the tie-break: greatest city number wins.
        best_city = -1
        best_count = INF
        for i in range(n):
            count = counts[i]
            if count < best_count or (count == best_count and i > best_city):
                best_city, best_count = i, count
        return best_city

File: problems-adapt/1301-1400/1334_most-isolated-node-under-a-reach-budget/solution_dijkstra_every_node.ts
function findMostIsolated(n: number, edges: number[][], budget: number): number {
    // Mirror each undirected edge both ways, so every node can run its own
    // Dijkstra over the adjacency list and pay only for real edges.
    const adj: number[][][] = Array.from({ length: n }, () => []);
    for (const [a, b, w] of edges) {
        adj[a].push([b, w]);
        adj[b].push([a, w]);
    }
    const INF = Infinity;

    // Min-heap of [d, node]
    const heap: number[][] = [];
    const push = (item: number[]): void => {
        heap.push(item);
        let i = heap.length - 1;
        while (i > 0) {
            const p = (i - 1) >> 1;
            if (heap[p][0] <= heap[i][0]) break;
            [heap[p], heap[i]] = [heap[i], heap[p]];
            i = p;
        }
    };
    const pop = (): number[] => {
        const top = heap[0];
        const last = heap.pop() as number[];
        if (heap.length > 0) {
            heap[0] = last;
            let i = 0;
            for (;;) {
                const l = 2 * i + 1;
                const r = l + 1;
                let m = i;
                if (l < heap.length && heap[l][0] < heap[m][0]) m = l;
                if (r < heap.length && heap[r][0] < heap[m][0]) m = r;
                if (m === i) break;
                [heap[m], heap[i]] = [heap[i], heap[m]];
                i = m;
            }
        }
        return top;
    };

    const counts: number[] = new Array(n);
    for (let src = 0; src < n; src++) {
        // Dijkstra from src: with positive weights the smallest tentative pop
        // is already final, so every node settles exactly once.
        const dist: number[] = new Array(n).fill(INF);
        dist[src] = 0;
        push([0, src]);
        while (heap.length > 0) {
            const [d, u] = pop();
            // Stale-entry guard: skip outdated heap records.
            if (d > dist[u]) {
                continue;
            }
            for (const [v, w] of adj[u]) {
                const nd = d + w;
                // Relax only when the route strictly improves.
                if (nd < dist[v]) {
                    dist[v] = nd;
                    push([nd, v]);
                }
            }
        }
        let count = 0;
        for (let v = 0; v < n; v++) {
            if (v !== src && dist[v] <= budget) {
                count++;
            }
        }
        counts[src] = count;
    }
    // Ascending scan with a strictly-smaller count (or equal count at a
    // larger index) implements the tie-break: greatest city number wins.
    let bestCity = -1;
    let bestCount = INF;
    for (let i = 0; i < n; i++) {
        const count = counts[i];
        if (count < bestCount || (count === bestCount && i > bestCity)) {
            bestCity = i;
            bestCount = count;
        }
    }
    return bestCity;
}
